opal/mca/btl/base/btl_base_am_rdma.c

```diff
@@ -795,69 +795,168 @@ static int mca_btl_base_am_rdma_progress (void)
 }
 
 static int mca_btl_base_am_atomic_64 (int64_t *operand, opal_atomic_int64_t *addr,
-                                      mca_btl_base_atomic_op_t op)
+                                      mca_btl_base_atomic_op_t op, opal_mutex_t *lock)
 {
     int64_t result = 0;
 
-    switch (op) {
-    case MCA_BTL_ATOMIC_ADD:
-        result = opal_atomic_fetch_add_64 (addr, *operand);
-        break;
-    case MCA_BTL_ATOMIC_AND:
-        result = opal_atomic_fetch_and_64 (addr, *operand);
-        break;
-    case MCA_BTL_ATOMIC_OR:
-        result = opal_atomic_fetch_or_64 (addr, *operand);
-        break;
-    case MCA_BTL_ATOMIC_XOR:
-        result = opal_atomic_fetch_xor_64 (addr, *operand);
-        break;
-    case MCA_BTL_ATOMIC_SWAP:
-        result = opal_atomic_swap_64 (addr, *operand);
-        break;
-    case MCA_BTL_ATOMIC_MIN:
-        result = opal_atomic_fetch_min_64 (addr, *operand);
-        break;
-    case MCA_BTL_ATOMIC_MAX:
-        result = opal_atomic_fetch_max_64 (addr, *operand);
-        break;
-    default:
-        return OPAL_ERR_BAD_PARAM;
+    if(OPAL_LIKELY(0 == ((((intptr_t) addr) & 0x7)))) { 
+        switch (op) {
```

Comment:
Not sure I am comfortable with this change. It is my understanding that MPI_Accumulate, etc the MPI implementation is allowed to require (or is required to) only operated on naturally aligned types. @bosilca Is this correct or is the user allowed to assume they can do atomics on byte-aligned data?

The problem with this change is that you essentially would need to disable the memory atomics interoperability because this lock is only in the BTL. That would break a lot of assumptions.

Reply:
For what it's worth this test passes with MPICH and Spectrum MPI (IBM mpi).

Reply:
My understanding is that the MPI standard currently does not mandate any alignment requirements for user buffers: https://github.com/mpiwg-rma/rma-issues/issues/3

Reply:
@devreal Buffers, yes. But if the user has an int64_t in the buffer it must be naturally aligned. Or that is what I would assume given that the MPI implementation may not be able to operate on them otherwise.

Reply:
Writing my own mini test program (no mpi), the memcopy method works on Power9. The data is correct, albeit misaligned. 

Now why users want to do this, I don't know. But I think it's valid.

Reply:
I agree that common sense dictates that it should be naturally aligned but there is nothing in the standard requiring that, AFAICS. We might want to fix that for 4.1 in the standard. I tried to come up with some wording in the ticket referenced above but this might well go beyond the RMA chapter...

Reply:
@awlauria Let's get confirmation that it should work. It is really silly to support something that is not recommended on a platform if we don't have to :).

Reply:
@devreal Maybe look at the datatype chapter. There is some language about alignment in that chapter. Not sure if any of it is relevant. Thats why I would like @bosilca to comment.

Reply:
@hjelmn I agree, but the optics are bad that OMPI segv's where other implementations don't. 

Even if we don't want to support this, maybe we can add a check at a higher level to not segv? Not sure off the top of my head where that can be prevented.

Reply:
Yeah. This is really messy. What I am worried about is silent wrong answers. Crashing is better there. Let me think about the whole stack that leads to this. Maybe this will be ok.

Reply:
> I don't think we should provide more atomicity than available on the system. Thus, if the windows alignment is 1, the safest approach is to force the lock approach, while for matching alignments we can use the atomic implementation, and error out if the user-provided pointer is misaligned.

That sounds good to me. If the user creates the window with matching alignment they would have to try extra hard to misalign the pointer, so letting it segfault seems appropriate.